### PR TITLE
Add shrinking to char arbitraries

### PR DESCRIPTION
**Pull request by jmid**
This PR adds shrinking to `char* arbitrary` generators.
In order to avoid shrinking a `numeral` char `'7'` into a non-numeral `char` it includes a dedicated shrinker.
The `printable` case is not broken by the existing `Shrink.char` shrinker.

To eat my own medicine, I then proceeded to add Generator, Shrinker, and Statistics tests
...only to find bugs :open_mouth: 
- it turns out the specification for `QCheck.Gen.printable` was wrong - it is nice to test it
- the distribution for `QCheck2.Gen.printable` revealed an off-by-one error in its internal `range` function, meaning that it would not produce, e.g., 'z'

### CODE REVIEW

**Comment by jmid** [comment]
I've refined the shrinker tests a bit to better exercise the expected behaviour.
In doing so, I noticed that in the range '!'-'&' (character codes 32-38) `QCheck2.printable` shrinks towards '!' (lower character codes) contradicting the documentation of shrinking towards 'a' (97). I therefore updated the documentation to match the behaviour better. I admit it is underspecified - but fully specifying the behaviour becomes unreadable, IMO.

Also updated the CHANGELOG, so this should be good to go.

**Comment by jmid** [comment]
@vch9 Can you take a quick look at this since it finds and fixes an issue with the `QCheck2.printable` generator? :pray:

File: src/core/QCheck2.mli [hunk 1]
```diff
@@ -280,7 +280,7 @@ module Gen : sig
     - [32] to [126], inclusive
     - ['\n']
 
-    Shrinks towards ['a'].
+    Shrinks towards ['a'] or lower character codes.
```
[the remark this answers is not in the document]
Comment (jmid) [reply src/core/QCheck2.mli:283]: The `or` is my lukewarm attempt at a more correct under-specification of a somewhat messy algorithmic behaviour:

https://github.com/c-cube/qcheck/blob/f09f42d02ef8fc25324845b3c5b2d90a542506fe/src/core/QCheck2.ml#L664-L668

It shrinks towards the front of that `printable_chars` list:
- `lower_alphabet;` (code 97-122)
- `before_lower_alphabet;` (32-96)
- `after_lower_alphabet;` (123-126)
- `newline` (code 10)

Here's the value of `printable_chars` before the fix (note the missing `z`, backquote, and `~`):
```ocaml
val printable_chars : char list =
  ['a'; 'b'; 'c'; 'd'; 'e'; 'f'; 'g'; 'h'; 'i'; 'j'; 'k'; 'l'; 'm'; 'n'; 'o';
   'p'; 'q'; 'r'; 's'; 't'; 'u'; 'v'; 'w'; 'x'; 'y'; ' '; '!'; '"'; '#'; '$';
   '%'; '&'; '\''; '('; ')'; '*'; '+'; ','; '-'; '.'; '/'; '0'; '1'; '2';
   '3'; '4'; '5'; '6'; '7'; '8'; '9'; ':'; ';'; '<'; '='; '>'; '?'; '@'; 'A';
   'B'; 'C'; 'D'; 'E'; 'F'; 'G'; 'H'; 'I'; 'J'; 'K'; 'L'; 'M'; 'N'; 'O'; 'P';
   'Q'; 'R'; 'S'; 'T'; 'U'; 'V'; 'W'; 'X'; 'Y'; 'Z'; '['; '\\'; ']'; '^';
   '_'; '{'; '|'; '}'; '\n']
```
Simply put, it shrinks towards earlier entries in this list.
That behaviour may change it the future however - so I don't think we should commit too much to this behaviour.

The first entry 'a' (code 97) is high (not smallest) - but most easy on the eye for us humans.
Concretely, if the failing `char` is `newline` with code 10 (smallest printable char code) it is guaranteed
to shrink to something higher. Other chars, e.g., `Z` (code 90) will shrink to a mix (both lower, e.g., 'A' and higher, e.g. 'a').

The test `printable_is_never_sign` shows how QCheck(1) will shrink towards char code 97 and thus `&` in the sequence
`!"#$%&` whereas QCheck2 will shrink towards `!`.

It is a mess!

**Comment by jmid** [comment]
@vch9's comment made me realize that reusing the `char` shrinker for `Gen.printable` in QCheck(1)
would break for `'\n'` (char code 10) by shrinking to a non-printable character.
I've therefore committed an explicit `Shrink.char_printable` shrinker along with a few unit tests of the behaviour.

Looking at the code for `Gen.printable` in QCheck.ml I see the same approach of QCheck2 using a local array/table.
This makes me think that the right solution(tm) would be to shrink towards the `'a'` entry in this array (at index 97 - 32 or so).
This could be done for both QCheck and QCheck2 and would clear up the weird QCheck2 character code jumping that is hard to specify.

File: src/core/QCheck2.mli (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmid) [reply src/core/QCheck2.mli]: Commit 8e63cb4 now simplifies the `Gen.printable` definition so we indeed just shrink towards 'a'.

**Comment by jmid** [comment]
I've now simplified the definition of `QCheck2.Gen.printable`. 
This makes the behaviour simpler to specify, as we always shrink towards 'a'.
Bonuses:
- we don't need a list of `printable_chars`.
- this cuts 0.5 sec of the QCheck2 shrinker benchmark and reduces major GC runs slightly from 368 to 364.

I've also updated the tests and the CHANGELOG accordingly, so this should be good to go.

**Comment by jmid** [comment]
Rebased on latest master to resolve expect-test output conflicts
